mlflow/store/tracking/rest_store.py

```diff
@@ -312,8 +335,47 @@ def get_experiment_by_name(self, experiment_name):
                 return None
             elif e.error_code == databricks_pb2.ErrorCode.Name(databricks_pb2.ENDPOINT_NOT_FOUND):
                 # Fall back to using ListExperiments-based implementation.
-                for experiment in self.list_experiments(ViewType.ALL):
-                    if experiment.name == experiment_name:
-                        return experiment
+                for experiments in self._paginate_list_experiments(ViewType.ALL):
+                    for experiment in experiments:
+                        if experiment.name == experiment_name:
+                            return experiment
                 return None
             raise e
+
+    def _paginate_list_experiments(self, view_type):
+        page_token = None
+        while True:
+            experiments = self.list_experiments(view_type=view_type, page_token=page_token)
+            yield experiments
+
+            if not experiments.page_token:
+                break
+            page_token = experiments.page_token
+
+    def list_experiments(
+        self,
+        view_type=ViewType.ACTIVE_ONLY,
+        max_results=_LIST_EXPERIMENTS_MAX_RESULTS_DEFAULT_IN_DATABRICKS,
+        page_token=None,
+    ):
+        # When we call `mlflow.tracking.MlflowClient.list_experiments` without specifying
+        # `max_results`, this function is called `max_results = None`:
+        #
+        # improt mlflow
+        # client = mlflow.tracking.MlflowClient(tracking_uri="https://...")
+        # client.list_experiments()
+        # ^ This calls `client._tracking_client.store.list_experiments(..., max_results=None)`
+        #
+        # To guarantee that the `max_results` field is always present in the request sent by
+        # MLflow client >= 1.18.0, set the default value if `max_results` is None.
+        # This allows MLflow backend in Databricks to infer the client version by examining
+        # the presence of the `max_results` field and determine what to return.
```

Comment:
Setting max_results in the default value to some number already effectively enforce a limit. What's the downside to allow user to explicitly specify None here? I assume the user would want an infinite limit effectively. If the backend just respect that and return min(Infinite, BackendInternalLimit), it should still work, no?

Reply:
If we allow `None` here, that means the new client also sends a request to the MLflow service backend in Databricks and can't distinguish the new and old clients.

Reply:
I think the point is to have a consistent behavior when observing certain input, irrespective of where it's coming from. The contract established here could be,
1) max_results cannot be 0 or negative. An error is returned;
2) max_results can be None / unset. The server will attempt to return all experiments. However, it could apply its internal limit if there are too many to return.
3) max_results is positive. In this case, the server will respect the input value. However, it could apply its internal limit if the requested value is too large.

The contract does not really care about where the request is sent from. Specifically on 2), it could be from an old client, which doesn't know max_results (hence unset), or from a new client, which knows how to use max_results but deliberately unsets the field to request maximum amount allowed by the server. Either way, the server only needs to check the field, and proceeds with 2) if max_results field does not exist, as listed in the contract.

Does it make sense?